Implement a controllable pRNG

konnov opened the pull request:
Closes #750. Closes #566. In this PR, we adopt a purely functional pseudo-random number generator. The benefits of this approach are as follows:

- [x] We are in full control of the random number generator. We can save the state and replay it at any point in time. For instance, the simulator saves the seed for the best trace, so it can be found on the first try the next time.
- [x] Our interface supports big integers, in contrast to `number` of the standard RNGs. This closes #566.
- [x] We can check whether a test used randomization by comparing the RNG state before and after running the test. This paves the way for #678. Postponed for a follow-up PR.
- [x] By printing the seed, we let the user to replay tests and random runs in greater detail later, e.g., with `--verbosity=4`.
- [x] Tests added for any new code
- [x] Entries added to the respective `CHANGELOG.md` for any new functionality

shonfeder commented on quint/test/rng.test.ts:
```diff
@@ -0,0 +1,55 @@
+    testForRange(newRng(), 2n ** 96n + 1n, 10000)
+  })
+
+  it('produces big integers in [0, over 2^96))', () => {
+    testForRange(newRng(), 2n ** 96n + 3n * 5n * 7n * 11n, 10000)
+  })
+
+  it('produces big integers in [0, over 2^256))', () => {
+    testForRange(newRng(), 2n ** 256n + 2n ** 33n + 11n, 10000)
+  })
```
Can you please also add a test showing that `n` calls starting from the same seed also produce the same set of sequences and states?

konnov commented on quint/src/rng.ts:
```diff
@@ -0,0 +1,155 @@
+
+/**
+ * Create a new instance of Rng, initialized from the current time and
+ * other system state. If you want to initialize the random number generator
+ * with a seed, create a new instance first, and then call `setState`.
```
Fixing this.

konnov commented on quint/src/rng.ts:
```diff
@@ -0,0 +1,155 @@
+  // Produce a random integer with the system RNG.
+  // Since the system generator is using `number`,
+  // the number is in the range of [0, 2^53).
+  let state: bigint =
+    BigInt(Math.floor(Math.random() * Number.MAX_SAFE_INTEGER))
+  return {
+    getState: (): string => {
+      return '0x' + state.toString(16)
+    },
```
yep, I did not want to depend on the implementation of another library. Since we are you using our own implementation now, we can do that.

konnov commented on quint/src/rng.ts:
```diff
@@ -0,0 +1,155 @@
+    },
+
+    next: (bound: bigint): bigint => {
+      let input: bigint = bound > 0n ? bound : 1n
```
Yes, let's have an assert here

konnov commented on quint/src/cliCommands.ts:
```diff
@@ -530,12 +536,12 @@ function replacer(_key: String, value: any): any {
-    return seedrandom(seed)
-  } else {
-    return seedrandom()
+    r.setState(seed)
```
Fixed that

konnov commented:
@shonfeder I think I've addressed all of your comments. Maybe we will extract this into a package one. I think this implementation is not very fast, since we do it over big integers, using plenty of divisions and modulo. Normally, people would be looking for an efficient implementation.

shonfeder reviewed (approved):
Awesome. Looks great. Thanks!

I've one concern still about the way the error is being essentially ignored when an invalid seed is set. I foresee lost time if we don't handle the error, but I won't block on that worry.

konnov commented on quint/src/cliCommands.ts:
```diff
@@ -532,12 +541,20 @@ function replacer(_key: String, value: any): any {
+      try {
+      seed = BigInt(seedText)
+    } catch (SyntaxError) {
+      console.error(`--seed must be a big integer, found: ${seedText}`)
```
This makes sense. I will fix that.

konnov commented on quint/src/cliCommands.ts:
```diff
@@ -355,6 +359,8 @@ export function runTests(prev: TypecheckedStage): CLIProcedure<TestedStage> {
             }
           }
+          // output the seed
+          out(chalk.gray(`    Use --seed=0x${testResult.seed.toString(16)} --match=${testResult.name} to repeat.`))
```
It's a hex representation of a big integer. The reason is that hex is more compact and looks cooler :)
